[tests] Refactor importmulti tests

jnewbery opened the pull request:
https://github.com/bitcoin/bitcoin/pull/14565 needs test coverage. This PR refactors wallet_importmulti.py to the following pattern:

1. Add `get_key()` and `get_multisig()` methods, which generate keys on node0 and return the priv/pubkeys and all scriptPubKey and address variants.
2. Add `test_importmulti()` method, which takes an importmulti request, sends it to node1 and tests against success and error codes/messages.
3. Add `test_address()` method, which takes an address, sends it as a getaddressinfo request to node1 and tests the values returned.

This does not add any specific testing for #14565, but makes it very straightforward to add that testing: `test_importmulti()` can be easily updated to test for returned warnings, and `test_address()` can be called multiple times against the different address variants for a singlesig/multisig.

jnewbery committed:
[tests] fix flake8 warnings in wallet_importmulti.py

jnewbery committed:
[tests] tidy up imports in wallet_importmulti.py

jnewbery committed:
[tests] add get_multisig function to wallet_importmulti.py

Adds a new get_multisig function which generates
a new multisig and returns the public keys,
private keys and all script and address types.

jnewbery committed:
[tests] add get_key function to wallet_importmulti.py

Adds a new get_key function which generates
a new key and returns the public key,
private key and all script and address types.

jnewbery committed:
[tests] add test_importmulti method to wallet_import.py

Adds a new test_importmulti method for testing the
importmulti RPC method.

jnewbery committed:
[tests] add test_address method to wallet_import.py

Adds a new test_address method for testing the
imported addresses.

jnewbery commented:
I've tidied up the intermediate commits and force-pushed. Should be ready for review.

MeshCollider commented:
Concept ACK, will review very soon

jnewbery committed:
[tests] Add docstring for wallet_importmulti.py

Adds a docstring describing the new importmulti test.

practicalswift commented on test/functional/wallet_importmulti.py:
```diff
@@ -32,595 +80,523 @@ def skip_test_if_missing_module(self):
+        addr_info = self.nodes[1].getaddressinfo(address)
+        for key, value in kwargs.items():
+            if addr_info[key] != value:
+                raise(AssertionError("key {} value {} did not match expected value {}".format(key, addr_info[key], value)))
```
Nit: `raise AssertionError(…)` is more idiomatic and consistent with the rest of the code base. Drop the unnecessary parens :-)

practicalswift commented on test/functional/wallet_importmulti.py:
```diff
@@ -32,595 +80,523 @@ def skip_test_if_missing_module(self):
+        Returns a named tuple of privkeys, pubkeys and all address and scripts."""
+        addrs = []
+        pubkeys = []
+        for i in range(3):
```
Nit: Use `_` instead of `i` to show that the variable is intentionally unused.

jnewbery commented:
@practicalswift nits fixed and redundant code removed: https://github.com/bitcoin/bitcoin/compare/6c76af02e8188d0dc66c315e4ecd80308e2fa0a1..12e59a60019732fa228872848121cca4431e7513

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -2,7 +2,18 @@
+
+- `get_key()` and `get_multisig()` are called to generate keys on node0 and
+  return the privkeys, pubkeys and all variants of scriptPubKey and address.
+- `test_importmulti() is called to send an importmulti call to node1, test
```
missing `

jnewbery replied:
fixed

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -209,10 +215,10 @@ def run_test(self):
-        assert_equal(address_assert['iswatchonly'], True)
-        assert_equal(address_assert['ismine'], False)
-        assert_equal(address_assert['timestamp'], timestamp)
+        self.test_address(address,
```
This should also test for solvability

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -511,18 +509,18 @@ def run_test(self):
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['solvable'], False)
+        self.test_address(multisig.p2sh_addr,
+                          solvable=False)
```
This should test watchonly=True

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -304,10 +308,9 @@ def run_test(self):
-        address_assert = self.nodes[1].getaddressinfo(address)
-        assert_equal(address_assert['iswatchonly'], False)
-        assert_equal(address_assert['ismine'], False)
-        assert_equal('timestamp' in address_assert, False)
```
n.b. we've lost this test now, but that seems fine

jnewbery replied:
I didn't think this test was important, so I removed it, but I've now readded it.

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -340,8 +343,7 @@ def run_test(self):
                               True)
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['timestamp'], timestamp)
+        self.test_address(multisig.p2sh_addr, timestamp=timestamp)
```
IMO it wouldn't hurt to test for solvability here too to ensure the same output as listunspent below

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -360,8 +362,8 @@ def run_test(self):
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['timestamp'], timestamp)
+        self.test_address(multisig.p2sh_addr,
+                          timestamp=timestamp)
```
Same here

MeshCollider commented on test/functional/wallet_importmulti.py:
```diff
@@ -579,8 +577,8 @@ def run_test(self):
-        address_assert = self.nodes[1].getaddressinfo(multisig.p2sh_addr)
-        assert_equal(address_assert['solvable'], True)
+        self.test_address(address,
+                          solvable=True)
```
test !ismine here?

meshcollider reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/14886/commits/12e59a60019732fa228872848121cca4431e7513, very nice, I've left a couple of minor comments inline but I assume you might want to keep this as a refactor rather than adding any extra testing here, so feel free to ignore the last few comments.

jnewbery commented on test/functional/wallet_importmulti.py:
```diff
@@ -55,20 +93,22 @@ def run_test(self):
         self.log.info("Should import an address")
-        address = self.nodes[0].getaddressinfo(self.nodes[0].getnewaddress())
+        key = self.get_key()
+        key = self.get_key()
```
oops. Fixed intermediate commit.

jnewbery commented on test/functional/wallet_importmulti.py:
```diff
@@ -84,24 +124,26 @@ def run_test(self):
         self.log.info("Should not allow a label to be specified when internal is true")
-        address = self.nodes[0].getaddressinfo(self.nodes[0].getnewaddress())
+        key = self.get_key()
+        address = key.p2pkh_addr
```
yes, removed

jnewbery commented:
Thanks @MeshCollider . I've addressed your review comments and pushed: https://github.com/bitcoin/bitcoin/compare/12e59a60019732fa228872848121cca4431e7513..ee3b21dccbeb0f9d4e99de869dbfaf625c159c7f

> I've left a couple of minor comments inline but I assume you might want to keep this as a refactor rather than adding any extra testing here, so feel free to ignore the last few comments.

Yes, the intent was for this to be a straight refactor. I think more comprehensive testing can be added as part of #14565. In fact, I've started doing that here: https://github.com/jnewbery/bitcoin/tree/pr14565.tests. We should go through that commit and make sure that it's comprehensively testing all return values.

ryanofsky reviewed (approved):
utACK ee3b21dccbeb0f9d4e99de869dbfaf625c159c7f. This is a test-only change and nice cleanup, and I think it should be merged as is. There are more tests in John's branch that depend on this, and #14565 will depend on those tests, and there are several other prs based on #14565, so this should be given some priority.

MarcoFalke commented on test/functional/wallet_importmulti.py:
```diff
@@ -108,6 +108,14 @@ def get_multisig(self):
                         CScript([OP_HASH160, witness_script, OP_EQUAL]).hex(),  # p2sh-p2wsh
                         script_to_p2sh_p2wsh(script_code))  # p2sh-p2wsh addr
 
+    def test_importmulti(self, req, success, error_code=None, error_message=None):
```
Agree, that at a minimum named arguments should be used at the caller.

Sjors commented:
Post merge ACK

jnewbery commented:
Thanks all for the review.

@ryanofsky - your review comments are fixed here: https://github.com/bitcoin/bitcoin/pull/14952
